OrbitCore/UserDefinedCaptureDataTest.cpp

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include <gtest/gtest.h>
+
+#include <string>
+
+#include "UserDefinedCaptureData.h"
+#include "capture_data.pb.h"
+
+using orbit_client_protos::FunctionInfo;
+
+FunctionInfo CreateFunctionInfo(const std::string& function_name, uint64_t function_address) {
+  FunctionInfo info;
+  info.set_name(function_name);
+  info.set_pretty_name(function_name);
+  info.set_loaded_module_path("/path/to/module");
+  info.set_address(function_address);
+  info.set_size(16);
+  info.set_file("file.cpp");
+  info.set_line(13);
+  return info;
+}
+
+TEST(UserDefinedCaptureData, InsertFrameTrack) {
+  UserDefinedCaptureData data;
+  FunctionInfo info = CreateFunctionInfo("fun0_name", 0);
+  data.InsertFrameTrack(info);
+  EXPECT_TRUE(data.ContainsFrameTrack(info));
+}
+
+TEST(UserDefinedCaptureData, InsertFrameTrackDuplicateFunctions) {
+  UserDefinedCaptureData data;
+  FunctionInfo info = CreateFunctionInfo("fun0_name", 0);
+  data.InsertFrameTrack(info);
+  data.InsertFrameTrack(info);
+  FunctionInfoSet set = data.frame_track_functions();
+  EXPECT_EQ(set.size(), 1);
+}
+
+TEST(UserDefinedCaptureData, InsertFrameTrackDifferentFunctions) {
+  UserDefinedCaptureData data;
+  FunctionInfo info0 = CreateFunctionInfo("fun0_name", 0);
+  FunctionInfo info1 = CreateFunctionInfo("fun1_name", 1);
+  data.InsertFrameTrack(info0);
+  data.InsertFrameTrack(info1);
+  FunctionInfoSet set = data.frame_track_functions();
+  EXPECT_EQ(set.size(), 2);
+}
+
+TEST(UserDefinedCaptureData, EraseFrameTrack) {
+  UserDefinedCaptureData data;
+  FunctionInfo info = CreateFunctionInfo("fun0_name", 0);
+  data.InsertFrameTrack(info);
+  data.EraseFrameTrack(info);
+  EXPECT_FALSE(data.ContainsFrameTrack(info));
+}
+
+TEST(UserDefinedCaptureData, EraseFrameTrackDuplicateFunctions) {
+  UserDefinedCaptureData data;
+  FunctionInfo info = CreateFunctionInfo("fun0_name", 0);
+  data.InsertFrameTrack(info);
+  data.InsertFrameTrack(info);
+  data.EraseFrameTrack(info);
+  EXPECT_FALSE(data.ContainsFrameTrack(info));
+}
+
+TEST(UserDefinedCaptureData, EraseFrameTrackDifferentFunctions) {
+  UserDefinedCaptureData data;
+  FunctionInfo info0 = CreateFunctionInfo("fun0_name", 0);
+  FunctionInfo info1 = CreateFunctionInfo("fun1_name", 1);
+  data.InsertFrameTrack(info0);
+  data.InsertFrameTrack(info1);
+  data.EraseFrameTrack(info0);
+  EXPECT_FALSE(data.ContainsFrameTrack(info0));
+  EXPECT_TRUE(data.ContainsFrameTrack(info1));
+}
+
+TEST(UserDefinedCaptureData, ContainsFrameTrackEmpty) {
+  UserDefinedCaptureData data;
+  FunctionInfo info = CreateFunctionInfo("fun1_name", 0);
+  EXPECT_FALSE(data.ContainsFrameTrack(info));
+}
+
+TEST(UserDefinedCaptureData, ContainsFrameTrackSingleFunction) {
```

Comment:
Isn't that the exact same test case as `InsertFrameTrack`. We might want to omit it (also following the recent talk about testing in our video hour).

Reply:
> also following the recent talk about testing in our video hour

I wasn't there to watch it :) What would be the reasoning here to omit this?

Reply:
Right, the idea was to prefer behavioural test cases over method/function base test cases.
The given example was a class simulating a glas, that can be filled and can you can drink from it.
So it has 4 methods: Constructor, Fill, Drink, IsFull.
Function based tests would be:
TestConstructor -> Create a Glass, we assume it is empty, but don't want to check its members (as this would not be a blackbox), so we use the "yet-untested" "IsFull" method.
TestFill -> Create a Glass, fill it, and again use the "IsFull" method, that does not really correspond to the test case.
TestDrink -> ...
TestIsFilled-> Now basically does not test anything new.

Instead it suggests behavioural based test cases:
TestAGlassCanBeCreated
TestAnEmptyGlassCanBeFilled
TestYouCanDrinkFromAFilledGlass

Reply:
FYI: https://youtu.be/FjwayiHNI1w?t=356

Reply:
But also note, this is super nitpicking. Feel free to ignore.

Reply:
Ok, thanks for the explanation. I went through the test cases and didn't really find one where that would make a big difference here, except for removing this test case here, which is obsolete and covered by the above behavior test (insertion causes contains to be true).